libvast/vast/span.hpp

```diff
@@ -3,573 +3,28 @@
-  template <class ExtentType>
-  class storage_type : public ExtentType {
-  public:
-    // KnownNotNull parameter is needed to remove unnecessary null check
-    // in subspans and constructors from arrays
-    template <class OtherExtentType>
-    constexpr storage_type(KnownNotNull data, OtherExtentType ext)
-      : ExtentType(ext), data_(data.p) {
-      // nop
-    }
-
-    template <class OtherExtentType>
-    constexpr storage_type(pointer data, OtherExtentType ext)
-      : ExtentType(ext), data_(data) {
-      // nop
-    }
-
-    [[nodiscard]] constexpr pointer data() const noexcept {
-      return data_;
-    }
-
-  private:
-    pointer data_;
-  };
-
-  storage_type<detail::extent_type<Extent>> storage_;
-
-  // The rest is needed to remove unnecessary null check
-  // in subspans and constructors from arrays
-  constexpr span(KnownNotNull ptr, index_type count) : storage_(ptr, count) {
-    // nop
-  }
-
-  template <size_t CallerExtent>
-  class subspan_selector {};
-
-  template <size_t CallerExtent>
-  [[nodiscard]] span<element_type, dynamic_extent>
-  make_subspan(index_type offset, index_type count,
-               subspan_selector<CallerExtent>) const {
-    span<element_type, dynamic_extent> tmp(*this);
-    return tmp.subspan(offset, count);
-  }
-
-  [[nodiscard]] span<element_type, dynamic_extent>
-  make_subspan(index_type offset, index_type count,
-               subspan_selector<dynamic_extent>) const {
-    VAST_ASSERT(size() - offset >= 0);
-    if (count == dynamic_extent)
-      return {KnownNotNull{data() + offset}, size() - offset};
-    VAST_ASSERT(size() - offset >= count);
-    return {KnownNotNull{data() + offset}, count};
-  }
-};
-
-// [span.comparison], span comparison operators
-template <class ElementType, size_t FirstExtent, size_t SecondExtent>
-constexpr bool operator==(span<ElementType, FirstExtent> l,
-                          span<ElementType, SecondExtent> r) {
-  return std::equal(l.begin(), l.end(), r.begin(), r.end());
-}
-
-template <class ElementType, size_t Extent>
-constexpr bool operator!=(span<ElementType, Extent> l,
-                          span<ElementType, Extent> r) {
-  return !(l == r);
-}
-
-template <class ElementType, size_t Extent>
-constexpr bool operator<(span<ElementType, Extent> l,
-                         span<ElementType, Extent> r) {
-  return std::lexicographical_compare(l.begin(), l.end(), r.begin(), r.end());
-}
-
-template <class ElementType, size_t Extent>
-constexpr bool operator<=(span<ElementType, Extent> l,
-                          span<ElementType, Extent> r) {
-  return !(l > r);
-}
-
-template <class ElementType, size_t Extent>
-constexpr bool operator>(span<ElementType, Extent> l,
-                         span<ElementType, Extent> r) {
-  return r < l;
+// TODO: Did not implement all comparison and relational operators as it is
+// tedious and not needed yet in this codebase.
+// See https://brevzin.github.io/c++/2020/03/30/span-comparisons/
```

Comment:
Just a note: we actually have a use case on the branch `topic/type-flatbuffers`, but that can use `std::lexicographical_compare` directly as well. It's also far from ready yet, so I'll refactor it when I get to it again.

Reply:
👍 sounds good. I recommend we add them as needed. As mentioned by Barry in the article from the comment, it's a bit tricky to get it right in all cases, so I'd rather not potentially introduce bugs in unused code right now.